cpp/dolfinx/mesh/graphbuild.cpp

```diff
@@ -22,124 +22,194 @@ namespace
 //-----------------------------------------------------------------------------
 // Compute local part of the dual graph, and return return (local_graph,
 // facet_cell_map, number of local edges in the graph (undirected)
-template <int N>
-std::pair<graph::AdjacencyList<std::int32_t>, std::vector<std::int64_t>>
+std::pair<graph::AdjacencyList<std::int32_t>, dolfinx::array2d<std::int64_t>>
 compute_local_dual_graph_keyed(
-    const graph::AdjacencyList<std::int64_t>& cell_vertices,
-    const mesh::CellType& cell_type)
+    const graph::AdjacencyList<std::int64_t>& cell_vertices, int tdim)
 {
   common::Timer timer("Compute local part of mesh dual graph");
 
-  const int tdim = mesh::cell_dim(cell_type);
   const std::int32_t num_local_cells = cell_vertices.num_nodes();
-  const int num_facets_per_cell = mesh::cell_num_entities(cell_type, tdim - 1);
-  const int num_vertices_per_facet
-      = mesh::num_cell_vertices(mesh::cell_entity_type(cell_type, tdim - 1));
+  if (num_local_cells == 0)
+  {
+    // Empty mesh on this process
+    dolfinx::array2d<std::int64_t> m(0, 0);
+    return {graph::AdjacencyList<std::int32_t>(0), m};
+  }
+
+  // Count number of cells with n vertices from 0-8
+  // covering interval(2) through to hex(8)
+  std::vector<int> count(9, 0);
+  for (int i = 0; i < num_local_cells; ++i)
+  {
+    assert(cell_vertices.num_links(i) < 9);
+    ++count[cell_vertices.num_links(i)];
+  }
+
+  int num_facets = 0;
+  int num_facet_vertices = 0;
+
+  // Map from number of cell vertices -> facet vertex list
```

Comment:
Explanation is a bit cryptic, would be better to expand. Start with high level intent with words.